include/albatross/src/models/patchwork_gp.hpp

```diff
@@ -0,0 +1,374 @@
+          typename BoundaryFeatureType>
+struct Fit<PatchworkGPFit<FitModel<ModelType, FitType>, GroupKey,
+                          BoundaryFeatureType>> {
+
+  using CovarianceRepresentation = decltype(FitType::train_covariance);
+
+  Grouped<GroupKey, FitModel<ModelType, FitType>> fit_models;
+  Grouped<GroupKey, Eigen::MatrixXd> information;
+  std::vector<BoundaryFeatureType> boundary_features;
+  // These matrices make up the block elements in Equation 5
+  Grouped<GroupKey, CovarianceRepresentation> C_dd;
+  Grouped<GroupKey, Eigen::MatrixXd> C_db;
+  Eigen::SerializableLDLT C_bb_ldlt;
+  Eigen::SerializableLDLT S_bb_ldlt;
+
+  Fit(){};
+
+  Fit(const Grouped<GroupKey, FitModel<ModelType, FitType>> &fit_models_)
+      : fit_models(fit_models_){};
+
+  Fit(const Grouped<GroupKey, FitModel<ModelType, FitType>> &fit_models_,
+      const Grouped<GroupKey, Eigen::MatrixXd> &information_,
+      const std::vector<BoundaryFeatureType> &boundary_features_,
+      const Grouped<GroupKey, CovarianceRepresentation> &C_dd_,
+      const Grouped<GroupKey, Eigen::MatrixXd> &C_db_,
+      const Eigen::SerializableLDLT &C_bb_ldlt_,
+      const Eigen::SerializableLDLT &S_bb_ldlt_)
+      : fit_models(fit_models_), information(information_),
+        boundary_features(boundary_features_), C_dd(C_dd_), C_db(C_db_),
+        C_bb_ldlt(C_bb_ldlt_), S_bb_ldlt(S_bb_ldlt_){};
+};
+
+template <typename BoundaryFunction, typename GroupKey>
+auto build_boundary_features(const BoundaryFunction &boundary_function,
+                             const std::vector<GroupKey> &groups) {
+  /*
+   * Loop through all combinations of groups (without permutations) and
+   * assemble the boundary features.
+   */
+  using BoundarySubFeatureType =
+      typename invoke_result<BoundaryFunction, GroupKey,
+                             GroupKey>::type::value_type;
+  using BoundaryFeatureType = BoundaryFeature<GroupKey, BoundarySubFeatureType>;
+
+  std::vector<BoundaryFeatureType> boundary_features;
+  for (std::size_t i = 0; i < groups.size(); ++i) {
+    for (std::size_t j = i + 1; j < groups.size(); ++j) {
+      const auto next_boundary_features = as_boundary_features(
+          groups[i], groups[j], boundary_function(groups[i], groups[j]));
+      if (next_boundary_features.size() > 0) {
+        boundary_features.insert(boundary_features.end(),
+                                 next_boundary_features.begin(),
+                                 next_boundary_features.end());
+      }
+    }
+  }
+  return boundary_features;
+}
+
+template <typename GroupKey, typename Solver, int Rows, int Cols>
+auto patchwork_solver(
+    const Grouped<GroupKey, Solver> &A,
+    const Grouped<GroupKey, Eigen::MatrixXd> &C,
+    const Eigen::SerializableLDLT &S,
+    const Grouped<GroupKey, Eigen::Matrix<double, Rows, Cols>> &rhs) {
+  // A^-1 rhs + A^-1 C (B - C^T A^-1 C)^-1 C^T A^-1 rhs
+  // A^-1 rhs + A^-1 C S^-1 C^T A^-1 rhs
+
+  const auto Ai_rhs = block_diag_solve(A, rhs);
+
```

Comment:
Does the rest of this function somehow implement the imposition of the boundary constraints?